Show SMS option when an account is using the SMS for 2fa

xknown opened the pull request:
#### Changes proposed in this Pull Request

This hides the Send Code by SMS buttons on the 2fa prompt as well as when one wants to disable 2fa for a given account.
<img width="426" alt="Screen Shot 2020-08-12 at 1 35 42 PM" src="https://user-images.githubusercontent.com/1689238/90266668-54222680-de22-11ea-8dd7-2d734d9cda0d.png">

<img width="735" alt="Screen Shot 2020-08-14 at 11 34 02 AM" src="https://user-images.githubusercontent.com/1689238/90266618-42d91a00-de22-11ea-9c7b-dbd70aef7855.png">

- Only show the sms button on the 2fa prompt (re-auth) and 2fa disable prompt when sms recovery phone number has been added
- Fix linter issues: remove extra space in translated text and remove `autoFocus` because it's an accessibility / usability issue to skip the warning text above the verification field. Also add id for the verification field to associate it with its label.

#### Testing instructions

- Log in to WP.com with an account with 2fa enabled via app, and with sms disabled (an a8c account for example).
- Navigate to http://calypso.localhost:3000/
- Remove the `twostep_auth` cookie from the `wordpress.com/public-api.wordpress.com` domain and refresh.
- Navigate to http://calypso.localhost:3000/me
- [ ] The 2fa re-auth prompt does NOT contain the "Send code via text message" button at the bottom
- Re-enter the verification code and go to two-step settings (me/security/two-step) and click "Disable two-step authentication"
- [ ] The 2fa disable prompt does NOT contain the "Send code via SMS" button at the bottom
- Repeat the above steps with an account that has 2fa enabled via app and an SMS recovery number added.
- [ ] The 2fa re-auth prompt does contain the "Send code via text message" button at the bottom
- [ ] The 2fa disable prompt does contain the "Send code via SMS" button at the bottom

Fixes #36809

xknown committed:
Only show the SMS button when the WP.com account is using the SMS for two factor

xknown commented:
Fixing the failing `lint-and-translate` test seems out of scope of this change and can likely be done on a different PR.

kwight commented:
@xknown From what I can tell, this has already been handled as part of https://github.com/Automattic/wp-calypso/pull/39515 – is that right?

xknown commented:
> @xknown From what I can tell, this has already been handled as part of #39515 – is that right?

Not really, the PR you mention only added support for webauthn, it only updated the place where the SMS button is shown, but it didn't change anything related to SMS.

kwight commented:
This will need a rebase and some attention then – I'll put it into the queue for janitorial rotation.

allilevine committed:
Only show the sms button when sms phone number has been added

wp-desktop reviewed (dismissed):
WordPress Desktop CI Failure for job "wp-desktop-mac".

@allilevine please inspect this job's build steps for breaking changes at [this link](https://circleci.com/gh/Automattic/wp-calypso/849831). For temporal failures, you may try to "Rerun Workflow from Failed".

Please also ensure this branch is rebased off latest Calypso.

allilevine commented on client/me/reauth-required/index.jsx:
```diff
@@ -268,7 +268,11 @@ const ReauthRequired = createReactClass( {
+					isSmsSupported={
+						method === 'sms' ||
+						( method === 'authenticator' &&
+							this.props.twoStepAuthorization.data.two_step_sms_last_four.length )
```
@xknown I added this condition to show the "Send code via text message" button only when an SMS number has been added. When there is no SMS second factor (like for an a8c account), there's no second method shown:
<img width="426" alt="Screen Shot 2020-08-12 at 1 35 42 PM" src="https://user-images.githubusercontent.com/1689238/90069229-042b4e80-dcc0-11ea-96a1-bcc9d8dae2d1.png">

Does that fix the issue while keeping the SMS fallback option around for people who enabled it?

nsakaimbo reviewed (dismissed):
👋 Does this PR affect any components that are displayed when logging out? For some reason desktop CI is consistently failing at the step:

```
1) Can logout from profile page
    - Can see app login page after logging out
```

([link](https://app.circleci.com/pipelines/github/Automattic/wp-calypso/70358/workflows/4129b7d5-ff37-49a6-a8c6-032967ac8579/jobs/849867))

CI is failing for two separate platforms (Mac and Linux) in the same location. If the desktop CI fails to pass even after a couple of re-tries you may want to take a closer look at any possible regressions this change might have introduced.

FWIW, it seems the last passing commit in this branch was b7b9183 "Only show the SMS button when the WP.com account is using the SMS for two factor". CI started failing with the subsequent commit.

allilevine committed:
Check that SMS number exists before checking length

allilevine committed:
Only show SMS option on 2-step disable prompt when an SMS recovery number has been added

Also fix linter issues: remove space from translated text and remove
autoFocus from verification code field. Add id that matches label for
attribute to verification code text field to improve accessibility.

allilevine commented:
Hi @vortfu -- would you be able to provide a Secops review for this PR?

xknown commented:
The change seems fine to fix the reported UI issue. However, I guess we still need to improve this in the future -- ideally we need to only allow the initial method a given user set up, the 2fa recovery number could be potentially implemented somewhere else.

allilevine commented:
> The change seems fine to fix the reported UI issue. However, I guess we still need to improve this in the future -- ideally we need to only allow the initial method a given user set up, the 2fa recovery number could be potentially implemented somewhere else.

Thanks! Do you mean that a 2fa recovery number could be added somewhere else and we wouldn't know about it? Or that the account recovery number should be completely separate from 2fa?

retrofox commented on client/me/reauth-required/index.jsx:
```diff
@@ -250,6 +250,9 @@ const ReauthRequired = createReactClass( {
 		const shouldEnableSmsButton =
 			this.state.smsRequestsAllowed || ( method === 'sms' && twoFactorAuthType === 'webauthn' );
 
+		const hasSmsRecoveryNumber = !! this.props.twoStepAuthorization.data.two_step_sms_last_four
```
Is it possible that some property there is undefined and it could trigger an Error exception? If so, maybe we could check the full path in order to avoid it. Something like the following:

```suggestion
		const hasSmsRecoveryNumber = !! this.props?.twoStepAuthorization?.data?.two_step_sms_last_four
```

allilevine replied:
Thanks, good idea! I made the change in both places.

retrofox commented on client/me/security-2fa-code-prompt/index.jsx:
```diff
@@ -192,6 +192,7 @@ class Security2faCodePrompt extends React.Component {
 
 	render() {
 		const method = twoStepAuthorization.isTwoStepSMSEnabled() ? 'sms' : 'app';
+		const hasSmsRecoveryNumber = !! twoStepAuthorization.data.two_step_sms_last_four?.length;
```
Similar suggestion to [this one](https://github.com/Automattic/wp-calypso/pull/36811/files#r495007558).

allilevine committed:
Check the full path when checking for the recovery number

retrofox reviewed (commented):
My tests

* Log in to WP.com with an account with 2fa enabled via app, and with sms disabled (an a8c account for example). ✅

<img src="https://user-images.githubusercontent.com/77539/94837155-416fab00-03ea-11eb-910e-8bf8f69a3d07.png" width="500px" />

<img src="https://user-images.githubusercontent.com/77539/94839740-ca3c1600-03ed-11eb-8c24-4f8ba8a033b8.png" width="500px" />

* Navigate to http://calypso.localhost:3000/ ✅

* Remove the twostep_auth cookie from the wordpress.com/public-api.wordpress.com domain and refresh. ✅
<img src="https://user-images.githubusercontent.com/77539/94845947-d1b3ed00-03f6-11eb-8f67-973c50a81f1b.png" width="500px" />

* Navigate to http://calypso.localhost:3000/me ✅ 

* The 2fa re-auth prompt does NOT contain the "Send code via text message" button at the bottom ✅ 

<img src="https://user-images.githubusercontent.com/77539/94846085-f4de9c80-03f6-11eb-9b63-dc1e044314c5.png" width="500px" />

* Go to two-step settings (me/security/two-step) and click "Disable two-step authentication"

It's like I can't authenticate from this point. The app asks, again and again, introduce the _the verification code generated by your authenticator app._

retrofox commented:
> It's like I can't authenticate from this point. The app asks, again and again, introduce the _the verification code generated by your authenticator app._

It happens in Chrome. Tested in Firefox and it works as expected.

<img src="https://user-images.githubusercontent.com/77539/94854372-2fe6cd00-0403-11eb-8ef8-6adc060ee4ee.png" width="500px" />

retrofox reviewed (approved):
LGTM. 👍

allilevine commented:
> > It's like I can't authenticate from this point. The app asks, again and again, introduce the _the verification code generated by your authenticator app._
> 
> It happens in Chrome. Tested in Firefox and it works as expected.
> 
> <img alt="" width="500px" src="https://user-images.githubusercontent.com/77539/94854372-2fe6cd00-0403-11eb-8ef8-6adc060ee4ee.png">

Verified that this is an issue on the master branch. Opened #46127
